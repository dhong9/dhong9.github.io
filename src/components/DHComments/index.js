import React, { useState } from "react";

// @mui material components
import Paper from "@mui/material/Paper";
import Typography from "@mui/material/Typography";
import Grid from "@mui/material/Grid";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";

// Sections components
import MKButton from "components/MKButton";
import MKInput from "components/MKInput";

// prop-types is a library for typechecking of props
import PropTypes from "prop-types";

import DHMarkdown from "components/DHMarkdown";

export default function DHComments({ comments }) {
  const [showReplyBox, setShowReplyBox] = useState(false);
  const [rootComment, setRootComment] = useState("");

  const handleReply = () => {
    setShowReplyBox(true);
  };

  const onAdd = () => {
    console.log(rootComment);
  };

  return (
    <Paper style={{ padding: "40px 20px" }}>
      {comments.map(({ id, body, name, create, isPlainText }) => (
        <Grid key={id} item xs={12} md={6}>
          <Card sx={{ display: "flex" }}>
            <CardContent sx={{ flex: 1 }}>
              <Typography component="h2" variant="h5">
                {name}
              </Typography>
              <Typography variant="subtitle1" color="text.secondary">
                {new Date(create).toLocaleString()}
              </Typography>
              <Typography variant="subtitle1" paragraph>
                {isPlainText ? body : <DHMarkdown>{body}</DHMarkdown>}
              </Typography>
              <Typography variant="a" href="#" color="primary" onClick={handleReply}>
                Reply...
              </Typography>
              {showReplyBox && (
                <>
                  <MKInput
                    variant="standard"
                    label="What do you think?"
                    placeholder="Add a comment"
                    InputLabelProps={{ shrink: true }}
                    multiline
                    fullWidth
                    rows={6}
                    onChange={(e) => setRootComment(e.target.value)}
                    value={rootComment}
                  />{" "}
                  <MKButton onClick={onAdd} type="submit" variant="gradient" color="info">
                    Add
                  </MKButton>
                </>
              )}
            </CardContent>
          </Card>
        </Grid>
      ))}
    </Paper>
  );
}

// Typechecking props of DHComments
DHComments.propTypes = {
  comments: PropTypes.arrayOf(PropTypes.object),
  addComment: PropTypes.func,
  pageName: PropTypes.string,
};

DHComments.defaultProps = {
  comments: null,
  addComment: null,
  pageName: "",
};
