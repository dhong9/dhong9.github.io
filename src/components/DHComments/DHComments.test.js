// React testing libraries
import React from "react";
import { render, fireEvent } from "@testing-library/react";

// Material Kit 2 React themes
import { ThemeProvider } from "@mui/material/styles";
import theme from "assets/theme";

// Component to test
import DHComments from "components/DHComments";

describe("DHComments", () => {
  let useEffect;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce((f) => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, "useEffect");
    mockUseEffect();
  });

  it("renders", () => {
    const comments = [
      {
        id: 1,
        project: 2,
        name: "John Adams",
        email: "john_adams@aol.com",
        body: "Play connect 4 with me",
        create: "2023-05-26T17:42:43.263337Z",
        updated: "2023-05-26T17:42:43.263383Z",
        active: true,
        parent: null,
      },
    ];

    const { container, getByText, getByPlaceholderText } = render(
      <ThemeProvider theme={theme}>
        <DHComments comments={comments} />
      </ThemeProvider>
    );

    // Click on the reply button
    const replyButton = getByText("Reply...");
    fireEvent.click(replyButton)

    // Write something in the reply box
    const commentBox = getByPlaceholderText("Add a comment");
    fireEvent.change(commentBox, { target: { value: "Connect 4 is fun!" } });

    expect(container).toMatchSnapshot()
  });
});
