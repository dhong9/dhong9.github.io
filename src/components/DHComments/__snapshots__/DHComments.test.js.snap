// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DHComments Adds a comment 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    style="padding: 40px 20px;"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6 css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1m5jx39-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-1p7cnz9-MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5 css-ozcfne-MuiTypography-root"
          >
            John Adams
          </h2>
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-1c7g210-MuiTypography-root"
          >
            5/26/2023, 10:42:43 AM
          </h6>
          <p
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-paragraph css-1mo0zj2-MuiTypography-root"
          >
            Play connect 4 with me
          </p>
          <span
            class="MuiTypography-root MuiTypography-a css-6tgqq3-MuiTypography-root"
            href="#"
          >
            Reply...
          </span>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1c3j9he-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-inw9yj-MuiInputBase-root-MuiInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-12marsf-MuiInputBase-input-MuiInput-input"
                id=":r0:"
                rows="6"
                style="height: 0px; overflow: hidden;"
              >
                Goodbye world!
              </textarea>
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-12marsf-MuiInputBase-input-MuiInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
            </div>
          </div>
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
          >
            <label
              class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-knlq9b-MuiFormControlLabel-root"
            >
              <span
                class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary css-xyabx4-MuiButtonBase-root-MuiCheckbox-root"
              >
                <input
                  class="PrivateSwitchBase-input css-1m9pwf3"
                  data-indeterminate="false"
                  type="checkbox"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17ceore-MuiSvgIcon-root"
                  data-testid="CheckBoxIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1dx2ob-MuiTypography-root"
              >
                Plain Text
              </span>
            </label>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1jnei4i-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Add
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
